File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.cs
public class Solution {
    public int MaxProfit(int[] prices) {
        int maxProfit = 0;
        int prev = int.MaxValue;
        foreach (int num in prices) {
            if (num > prev) {
                maxProfit += num - prev;
            }
            prev = num;
        }
        return maxProfit;
    }
}

File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:

        cache = {}
        
        def dp(i, canBuy):

            if i == len(prices):
                return 0

            if (i, canBuy) in cache:
                return cache[(i, canBuy)]
            
            #dont do anything
            p1 = dp(i + 1, canBuy)

            if canBuy:
                p2 = -prices[i] + dp(i + 1, not canBuy)
            else:
                p2 = prices[i] + dp(i + 1, not canBuy)
            
            cache[(i, canBuy)] = max(p1, p2)
            return cache[(i, canBuy)]
        
        return dp(0, True)
